botocore/translate.py

```diff
@@ -11,22 +11,46 @@
 # distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
 # ANY KIND, either express or implied. See the License for the specific
 # language governing permissions and limitations under the License.
+import copy
+
 from botocore.utils import merge_dicts
 
 
-def build_retry_config(endpoint_prefix, retry_model, definitions):
+def build_retry_config(endpoint_prefix, retry_model, definitions,
+                       client_retry_config=None):
     service_config = retry_model.get(endpoint_prefix, {})
     resolve_references(service_config, definitions)
     # We want to merge the global defaults with the service specific
     # defaults, with the service specific defaults taking precedence.
     # So we use the global defaults as the base.
-    final_retry_config = {'__default__': retry_model.get('__default__', {})}
+    #
+    # A deepcopy is done on the retry defaults because it ensures the
+    # retry model has no chance of getting mutated when the service specific
+    # configuration or client retry config is merged in.
+    final_retry_config = {
+        '__default__': copy.deepcopy(retry_model.get('__default__', {}))
```

Comment:
Yeah there was a bug. The only time it would be executed though is if you created a dynamodb client and then created other types of clients after as the other clients would inherit the 10 max_attempts of dynamodb. I added the `test_service_specific_defaults_do_not_clobber` to make sure that does not happen again.